### PR TITLE
perf number mild and extreme outlier detection 

### PR DESCRIPTION
**Pull request by carolynhu**
This experimental PR is using box plot method to calculate and detect outliers. Here is the graph:
Screen Shot 2019-12-23 at 3 32 54 PM

### CODE REVIEW

File: perf_dashboard/assets/js/alert_common_func.js [hunk 1]
```diff
@@ -26,7 +26,7 @@ window.generateChart = function(chartID, modesData) {
       }
     },
     axisY: {
-      title: "p90 Latency Pattern in milliseconds",
+      title: "Latency Pattern (milliseconds)",
```
Comment (howardjohn) [inline perf_dashboard/assets/js/alert_common_func.js]: "latency" is not meaningful without the percentile

File: perf_dashboard/helpers/download.py [hunk 2]
```diff
@@ -80,15 +80,15 @@ def download_benchmark_csv(days):
...
+        cur_release_dates[i] = [0] * 4
+        cur_release_dates[i] = [sub_str[0:4], sub_str[4:6], sub_str[6:8], sub_str]
 
     master_release_dates = [[]] * len(master_release_names)
     for i in range(len(master_release_names)):
         master_release = master_release_names[i]
         sub_str = master_release[len("master") + 1:].split("-")[0]
-        master_release_dates[i] = [0] * 3
-        master_release_dates[i] = [sub_str[0:4], sub_str[4:6], sub_str[6:8]]
+        master_release_dates[i] = [0] * 4
```
Comment (howardjohn) [inline perf_dashboard/helpers/download.py]: I have no clue what this does, should we make this more clear or add comments maybe?

File: perf_dashboard/helpers/outlier_detection.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+def find_outliers(pattern_data):
+    perf_val = []
+    for data in pattern_data:
+        if data[3] != 'null':
+            perf_val.append((data[3], data[4], data[5]))
```
Comment (howardjohn) [inline perf_dashboard/helpers/outlier_detection.py]: what is in 3, 4, 5? Should we make it an object so we can refer to it by data.foo, data.bar, etc?

File: perf_dashboard/helpers/outlier_detection.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+def calculate_fences(perf_val):
+    Q1, Q3, IQR = cacluate_interquartile_range(perf_val)
+    lower_inner_fence = Q1 - 1.5 * IQR
+    upper_inner_fence = Q3 + 1.5 * IQR
+    lower_outer_fence = Q1 - 3 * IQR
+    upper_outer_fence = Q3 + 3 * IQR
+    return lower_inner_fence, upper_inner_fence, lower_outer_fence, upper_outer_fence
+
+
+def cacluate_interquartile_range(perf_val):
```
Comment (howardjohn) [inline perf_dashboard/helpers/outlier_detection.py]: are there any python std library to do this sort of thing for us?

File: perf_dashboard/assets/js/alert_common_func.js (hunk 1 above)
Comment (carolynhu) [reply perf_dashboard/assets/js/alert_common_func.js]: the X-axis title with P90 or P99 is in specified in html file, here is just common js file which stands for the Y-axis title.

File: perf_dashboard/helpers/download.py (hunk 2 above)
Comment (carolynhu) [reply perf_dashboard/helpers/download.py]: This just parsing the release with SHA name. Sure, I will add a common or make it interpretable.

File: perf_dashboard/helpers/outlier_detection.py (hunk 3 above)
Comment (carolynhu) [reply perf_dashboard/helpers/outlier_detection.py]: yes, good catch.

File: perf_dashboard/helpers/outlier_detection.py (hunk 4 above)
Comment (carolynhu) [reply perf_dashboard/helpers/outlier_detection.py]: maybe have, but I think we may want more customized stuff later.

**Comment by mandarjog** [comment]
@carolynhu we will have two measurements in the system.
1. Goal: this is not data dependent. For examples 2ms of additional latency at p99.
    Or idealized `Goal == baseline` 
2. Trend analysis
    a. Discard or re-analyze outliers.
    b. Check if the stat is trending in the right direction.

This PR is a solution for 2a.

For 2b we just want a simple linear regression, and for 2a we want the distance from regression line to be subject to outlier detection, not the sample itself.

All this is predicated on having low variance data. If we need to produce some numbers for Mako, I am ok with some experimentation. However I do not think this is giving us the correct signal.

**Comment by carolynhu** [comment]
after synced with @mandarjog The correct solution should be applying outlier analysis based on trending data.

**Comment by carolynhu** [comment]
rebased with master

**Comment by carolynhu** [comment]
> So my understanding, we look at the last n (10?) results. If any of them are over `Q3 + 1.5 IQR`, we mark it as an outlier? So if the data was 1ms, 2ms, 3ms, 4ms, 5ms, 6ms, 7ms, ... we would never have an outlier, yet we are getting worse and worse over time? Is that right or am I misunderstanding this?
> 
> It seems what we are after is actually whether there is a statistically significant increase in latency, not if there is an outlier? what do you think?

Ok, let me explain the process:
(1) First, we get the (telemetry_mode - baseline) y-series data, here I set it to 40 days.
(2) Then, sort by perf_num in increasing order
(3) Calculate fences based on this sorted array
(4) According to the outliers definition of box-plot, mark those data which are not in the range of normal data as outliers.

So, next step is to waiting for our perf pipeline stable, applying this to the stable data, and calculate the initial thresholds per telemetry_mode.

I'm still in discussing with @mandarjog about how we want to get this done.

**Comment by carolynhu** [comment]
@mandarjog Synced with Ankit, we need to figure out which kind of parameters that Mako Regression Analyzer is expecting.

**Comment by carolynhu** [comment]
@mandarjog @howardjohn Design Proposal: Observations and Concerns of Istio Performance Pipeline Using Mako https://docs.google.com/document/d/1S3i8z20iVkge1MpuoQj86fCg54BjIbBTF7tAdWQltt8/edit#heading=h.xw1gqgyqs5b

**Comment by carolynhu** [comment]
Close this PR, use this ticket https://github.com/istio/istio/issues/24520 to track all the related works.

**Pull request closed without merging**